common/perf-test-core/src/main/java/com/azure/perf/test/core/BatchPerfTest.java

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.perf.test.core;
+
+import reactor.core.publisher.Mono;
+
+/**
+ * Represents the abstraction of a Performance test class running operations in batches.
+ *
+ * <p>
+ * The performance test class needs to extend this class. The test class should override {@link BatchPerfTest#runBatch()}
+ * and {@link BatchPerfTest#runBatchAsync()} methods and the synchronous and asynchronous test logic respectively.
+ * To add any test setup and logic the test class should override {@link BatchPerfTest#globalSetupAsync()}
+ * and {@link BatchPerfTest#globalCleanupAsync()} methods .
+ * </p>
+ * @param <TOptions> the options configured for the test.
+ */
+public abstract class BatchPerfTest<TOptions extends PerfStressOptions> extends ApiPerfTestBase<TOptions> {
```

Comment:
"Batch vs single" and "api-based vs event-based" are two different dimensions.

For api-based SDKs, we use `BatchPerfTest` for APIs which represent multiple logical operations, and `PerfTest` for APIs which represent a single logical operation.  These are different base classes because one exposes `abstract void Run()` and the other exposes `abstract int Run()`.  We could have exposed both methods on a single base class (as `virtual` rather than `abstract`), but I thought this might be confusing to users, especially since "batch" scenarios are much less common.

For event-based SDKs, the current design is to use one base class `EventPerfTest` for both "batch" and "single" scenarios.  In a "batch" scenario, the test might call `eventRaised()` multiple times in response to a single event raised by the SDK, but this seems fine.  Sample test code might look like this:

```c#
public void receiveMessageBatch(Message[] messages) {
    foreach (var message in messages) {
        // consume message body
        eventRaised();
    }
}
```

We could add a method to `EventPerfTest` like `eventsRaised(int events)`, which could change the test code to this, but I don't think it's worth it:

```c#
public void receiveMessageBatch(Message[] messages) {
    foreach (var message in messages) {
        // consume message body
    }
    eventsRaised(messages.length);
}
```

The reason we need two classes for api-based and one for event-based is due to calling pattern.  For api-based, the perf framework calls methods on the perf test class, so it's cleaner to expose only the methods needed on two different base classes.  For event-based, the perf test class calls back into the framework to report completed operations, so a single base class is sufficient, even if we decide to add `eventsRaised(int events)` in the future.

@srnagar: Do you have same code for batch event-based SDKs in Java?  How are the batch of messages raised in a single event?

Reply:
@mikeharder yes, we do have batch event-based APIs in Java and in other languages too. 
https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventProcessorClientBuilder.java#L574

Reply:
@srnagar: I agree we should test both the single-event and batch APIs in our SDK, since they may have different performance (and performance may vary by `maxBatchSize`).

However, I'm still not convinced the perf framework needs an `eventRaised(int events)` method, based on my arguments above.  I want to keep the perf framework as lightweight as possible, since whatever APIs we add need to be ported to every language.

I'm not opposed to adding `eventRaised(int events)` in the future, if we decide it's necessary for either convenience or because calling `eventRaised()` in a loop adds too much overhead.